### ISSUE TITLE
`drop_terminal_measurements` fails for measurements on dimension>2 Qids

### ISSUE THREAD

**Issue by richrines1** [quantumlib/Cirq#5994]
**Description of the issue**

the `cirq.drop_terminal_measurements` transformer fails if any measurements are on `Qid`s with dimension greater than 2

**How to reproduce the issue**

```python
_ = cirq.drop_terminal_measurements(cirq.Circuit(cirq.measure(cirq.LineQid(0, dimension=2))))  # ok
_ = cirq.drop_terminal_measurements(cirq.Circuit(cirq.measure(cirq.LineQid(0, dimension=3))))  # ValueError: Wrong shape of qids for <cirq.I>. Expected (2,) but got (3,) <[cirq.LineQid(0, dimension=3)]>.
```

**Cirq version**

```
1.2.0.dev20230105212249
```

**Comment by richrines1** [comment]
it looks like this arises when inserting [X/I operations according to `invert_mask`](https://github.com/quantumlib/Cirq/blob/a3520bc3570da018d1e89d2d788e889dcd0d6417/cirq-core/cirq/transformers/measurement_transformers.py#L295) - out of curiosity how should `invert_mask` be interpreted for qudit measurements?

**Comment by dstrain115** [comment]
Adding to triage discuss to answer above question.  How should we handle invert_mask when using qudits?

**Comment by daxfohl** [comment]
To be compatible with `deferred_measurement_transformer` it has to be a `mod_add(1)`. https://github.com/quantumlib/Cirq/blob/27dd6076f8fb9631c20d6234134edd76c9cdd681/cirq-core/cirq/transformers/measurement_transformers.py#L116

**Comment by daxfohl** [comment]
I take that back, the `deferred_measurement_transformer` code doesn't handle it at all. It'd just require changing `X` to `XPowGate` and explicitly passing in the dimension though. I think this is the best option nonetheless. If there's a desire to allow `int` in the `invert_mask`, for completions sake, that would translate to an exponent in `deferred_measurement_transformer`'s `XPowGate`.

**Comment by richrines1** [comment]
i think that's probably what i'd expect to happen, though also noticing that the simulator currently does something different (applying `^= 1` to the result iff the measured value is either 0 or 1)

also fwiw it looks like the same pattern comes up again in [`cirq.circuits.circuit._decompose_measurement_inversions`](https://github.com/quantumlib/Cirq/blob/27dd6076f8fb9631c20d6234134edd76c9cdd681/cirq-core/cirq/circuits/circuit.py#L2703), leading to e.g.,
```python
measurement = cirq.measure(cirq.LineQubit(0, dimension=3), invert_mask=(True,))
cirq.has_unitary(cirq.Circuit(measurement))  # ValueError: Wrong shape of qids for <cirq.X>. Expected (2,) but got (3,) <[cirq.LineQid(0, dimension=3)]>.
```

**Comment by tanujkhattar** [comment]
`invert_mask` is supposed to be a classical post processing on the measurement outcomes of a qubit which conveys that we should "flip" the measurement results. As discussed above, extending this to general `qids` would require us start supporting `add_mod(k)`, so for qutrits one could either add `1` or add `2` (both mod `3`) to the measurement results. 

However, note that we already support `confusion_map` parameter in `MeasurementGate`, which provides a more general way to specify what classical post processing should be applied to the measurement results. Note that `confuson_map` was introduced after `invert_mask` and `invert_mask` has been kept around for legacy and backwards compatibility. 

Going forward, I recommend we do the following:
1. Raise an error for general qudits if an `invert_mask` is specified. 
2. Use `confusion_map` to specify the classical post processing that should be applied to general qudits. 
3. Potentially deprecate `invert_mask` for next major release (cirq 2.0).


@daxfohl @richrines1 Let me know what you think.

**Comment by daxfohl** [comment]
Makes sense. invert_mask would be a horrible name for an inc_mod function, even though it's the only reasonable qudit behavior. More likely to confuse people than anything.

I think no need to deprecate it though, it's still a useful and meaningful shortcut.

**Comment by madhurimaparuchuri** [comment]
Hi, is anyone working on this issue, if not I will pick this

**Comment by daxfohl** [comment]
Apparently `invert_mask` does have a defined meaning for qudits, and it's not `inc_mod`: it swaps 0 and 1, and leaves other values the same. Code here https://github.com/quantumlib/cirq/blob/3fefe2984a1203c0bf647c1ea84f4882b05f8477/cirq-core/cirq/sim/simulator.py#L823 has existed since 2020. There are also unit tests testing that behavior. So, I think not worth a breaking change.

Thus, instead of changing the behavior of `invert_mask`, I created a PR that implements `drop_terminal_measurements` with this behavior above.
